dvc/repo/gc.py

```diff
@@ -82,23 +108,31 @@ def gc(  # noqa: PLR0913
                 jobs=jobs,
                 revs=[rev] if rev else None,
                 num=num or 1,
-            ).values():
-                used_obj_ids.update(obj_ids)
+            ).items():
+                odb_to_obj_ids.append((odb, obj_ids))
+
+    if not_in_remote:
+        used_obj_ids = _used_obj_ids_not_in_remote(self, remote, jobs, odb_to_obj_ids)
+    else:
+        used_obj_ids = set()
+        for _, obj_ids in odb_to_obj_ids:
+            used_obj_ids.update(obj_ids)
 
     for scheme, odb in self.cache.by_scheme():
```

Comment:
@efiop I might be missing something but `by_scheme` here returns a duplicated `odb` (for `repo` and `local` schemes) so we are doing a redundant `odb.all`.

What is the scenario where `repo` and `local` schemes might point to different `odb`?

Should we prevent the redundant call where both point to the same `odb`?